docs/userguide/apps.rst

```diff
@@ -125,10 +118,67 @@ to include attributes from the original function (e.g., its name).
     my_max_app = python_app(my_max)
 
 
-Apps created using ``python_app`` as a function will work just like those which use it as a decorator.
 
-Returns
-^^^^^^^
+Inputs and Outputs
+^^^^^^^^^^^^^^^^^^
+
+Python apps may be passed any Python type as an input and return any Python type, with a few exceptions.
+There are several classes of allowed types, each with different rules.
+
+- *Python Objects*: Any Python object that can be saved with
+  `pickle <https://docs.python.org/3/library/pickle.html>`_ or `dill <https://dill.readthedocs.io/>`_
+  can be used as an import or output.
+  All primitive types (e.g., floats, strings) are valid as are many complex types (e.g., numpy arrays).
+- *Files*: Pass files as inputs as a :py:class:`~parsl.data_provider.files.File` object.
+  Parsl can transfer them to a remote system and update the ``File`` object with a new path.
+  Access the new path with ``File.filepath`` attribute.
+
+  .. code-block:: python
+
+      @python_app
+      def read_first_line(x: File):
+          with open(x.filepath, 'r') as fp:
+              return fp.readline()
+
+  Files can also be outputs of a function, but only through the ``outputs`` kwargs (described below).
+- *Parsl Futures*. Functions can receive results from other Apps as Parsl ``Future`` objects.
+  Parsl will establish a dependency on the App(s) which created the Future(s)
+  and start executing as soon as the preceding ones complete.
+
+  .. code-block:: python
+
+    @python_app
+    def capitalize(x: str):
+        return x.upper()
+
+    input_file = File('text.txt')
+    first_line_future = read_first_line(input_file)
+    capital_future = capitalize(first_line_future)
+    print(capital_future.result())
+
+  See the section on `Futures <futures.html>`_ for more details.
+
+
+Learn more about the types of data allowed in `the data section <data.html>`_.
+
+Special Keyword Arguments
++++++++++++++++++++++++++
+
+Parsl apps can use the following special reserved keyword arguments:
```

Comment:
This can use an example because the uninitiated reader may not be familiar that you are referring to passing arguments _to the decorator_ rather than the underlying function.

Reply:
Absolutely, and I also a added a section which differentiates "function" and "decorator" kwargs.